### PR TITLE
ReadOnly Interface Dispatch using PIC Slots

### PR DESCRIPTION
**Pull request by r30shah**
Changes for supporting no patching Interface Dispatch sequence using PIC slots for read only code.
Updating the PIC Slots for Interface Dispatch call site in PICBuilder helper for Z works as following.

```
_interfaceCallHelperReadOnlyMultiSlots 
{
...
   currentOffset = 0
   checkNextSlot:
   CompareAndSwap defaultValueInClassField,J9Class,$(currentOffset+PICSlotsData)
   If CompareAndSwap is successfull
      Update the methodAddress field in the PIC Slot
      Jump doneCacheUpdate

   If CachedClassInPIC == J9Class
      Jump doneCacheUpdate

   currentOffset += sizeof single PIC slot
   if currentOffset <= lastOffset
      Jump checkNextSlot

   doneCacheUpdate:
...
}
```

At call site, we check the class field on each slot one after another to see if we get hit and branch to cached method address before falling to slow path and call the PIC builder helper.

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

### CODE REVIEW

**Comment by r30shah** [comment]
@fjeremic  Can I get a review on these changes. I have documented the approach I have taken to update the PIC slots for read only code in the commit message. Current approach is little bit different then the default JDK mode (Documented in https://github.com/eclipse/openj9/issues/10795) where in read only mode, I check all the slots when updating the cache slots, and if the I am updating the slot with the class already cached, I skip to the dispatch sequence.

**Comment by fjeremic** [comment]
Putting this on my TODO. It will take a day or so until I get to review this carefully.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp [hunk 1]
```diff
@@ -1653,3 +1653,67 @@ TR::S390VirtualUnresolvedReadOnlySnippet::getLength(int32_t estimatedSnippetStar
...
+   TR::Node *callNode = getNode();
+
+   TR::SymbolReference *interfaceCallReadOnlySymRef = cg()->symRefTab()->findOrCreateRuntimeHelper(TR_S390interfaceCallHelperMultiSlotsReadOnly,
+                                                                                                               false, false, false);
+   
+   TR::GlobalFunctionCallData dataDestination(interfaceCallReadOnlySymRef, callNode, cursor, cg(), self());
+
+   cursor = cg()->getObjFmt()->encodeGlobalFunctionCall(dataDestination);
+
+   uint8_t *helperCallRA = cursor;
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390J9CallSnippet.cpp:1670]: Why not just change the type of this variable to be `intptr_t` and remove the casts below?

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp [hunk 2]
```diff
@@ -1653,3 +1653,67 @@ TR::S390VirtualUnresolvedReadOnlySnippet::getLength(int32_t estimatedSnippetStar
...
+   {
+   uint8_t *bufferPos = snippet->getSnippetLabel()->getCodeLocation();
+   TR::SymbolReference *resolveVirtualDispatchReadOnlySymRef = _cg->getSymRef(TR_S390interfaceCallHelperMultiSlotsReadOnly);
+
+   TR::SymbolReference *methodSymRef = snippet->getNode()->getSymbolReference();
+   // TODO: Need the address of the ccGlobalFunctionCallData   
+   printSnippetLabel(pOutFile, snippet->getSnippetLabel(), bufferPos, "Interface Call Snippet using Multi Slots");
+   printPrefix(pOutFile, NULL, bufferPos, 6);
+   trfprintf(pOutFile, "LGRL \tGPR14, <%p>\t# Load address of the Helper Method, targetAddress = <%p>",
+                    (void *)*(int32_t*)(bufferPos+sizeof(int16_t)),
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390J9CallSnippet.cpp:1693]: `void *` vs `void*` inconsistent. Also need spaces between arithmetic `+` operators.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 3]
```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
...
+   // Intialization of CCData
+   ccInterfaceDataAddress->slot1Class  = 0;
+   ccInterfaceDataAddress->slot1Method = 0;
+   ccInterfaceDataAddress->slot2Class  = 0;
+   ccInterfaceDataAddress->slot2Method = 0;
+   ccInterfaceDataAddress->slot3Class  = 0;
+   ccInterfaceDataAddress->slot3Method = 0;
+
+   ccInterfaceDataAddress->cpAddress = reinterpret_cast<intptr_t>(callNode->getSymbolReference()->getOwningMethod(cg()->comp())->constantPool());
+   ccInterfaceDataAddress->cpIndex = (intptr_t) callNode->getSymbolReference()->getCPIndexForVM();
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1691]: `static_cast` needed here.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 4]
```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
...
+   ccInterfaceDataAddress->interfaceClass = 0;
+   ccInterfaceDataAddress->iTableIndex = 0;
+   ccInterfaceDataAddress->j2iThunkAddress = virtualThunk;
+   ccInterfaceDataAddress->totalSizeOfPicSlots = (3 /*Total Number of PICSlots*/ * sizeof(ccInterfaceDataAddress->slot1Class)) + (3 * sizeof(ccInterfaceDataAddress->slot1Method));
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1695]: Can we use a `static const` inside of `ccInterfaceData` to store the constant `3` and use it here instead? This seems quite error prone as the magic constant gets used several times in this function and we can replace it with a properly named and scoped constant that is defined right in the struct.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 5]
```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
...
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg());
+   cursor = generateRILInstruction(cg(), TR::InstOpCode::LARL, callNode, regRA, doneLabel, cursor);
+
+   TR::RegisterDependencyConditions *preDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(dependencies->getPreConditions(), NULL,
+                                                                                                         dependencies->getAddCursorForPre(), 0, cg());
+   
+   TR::RegisterDependencyConditions *postDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(dependencies, 0, numberOfAdditionalPostDeps, cg());
+   // Removing the pre dependencies from the postDeps
+   postDeps->setAddCursorForPre(0);
+   postDeps->setNumPreConditions(0, trMemory());
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1747]: Why is this needed?

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 6]
```diff
@@ -1649,6 +1659,140 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
...
+   TR::LabelSymbol *interfaceCallHelperSnippetLabel = generateLabelSymbol(cg());
+   TR::S390InterfaceCallReadOnlySnippet *snippet = new (trHeapMemory()) TR::S390InterfaceCallReadOnlySnippet(
+      cg(),
+      callNode,
+      interfaceCallHelperSnippetLabel,
+      doneLabel,
+      interfaceDataAddress);
+   // Cache miss, going to slow path to lookup the interface method.
+   cursor = new (trHeapMemory()) TR::S390RILInstruction(TR::InstOpCode::BRCL, callNode, 0xF, snippet, NULL, cg());
+   cursor = new (trHeapMemory()) TR::S390NOPInstruction(TR::InstOpCode::NOP, 2, callNode, cg());
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1782]: Why is the NOP needed?

File: runtime/compiler/z/runtime/PicBuilder.m4 [hunk 7]
```diff
@@ -155,6 +155,24 @@ SETVAL(eq_returnAddressDirectMethod,8)
...
+SETVAL(eq_InterfaceCallHelperClassPICSlot2,16)
+SETVAL(eq_InterfaceCallHelperMethodPICSlot2,24)
+SETVAL(eq_InterfaceCallHelperClassPICSlot3,32)
+SETVAL(eq_InterfaceCallHelperMethodPICSlot3,40)
+SETVAL(eq_InterfaceCallHelper_cpAddress,48)
+SETVAL(eq_InterfaceCallHelper_cpIndex,56)
+SETVAL(eq_InterfaceCallHelper_interfaceClass,64)
+SETVAL(eq_InterfaceCallHelper_methodIndex,72)
+SETVAL(eq_InterfaceCallHelperTotalPICSlotSize,88)
+SETVAL(eq_InterfaceCallHelper_isCacheFullFlag,96)
```
Comment (fjeremic) [inline runtime/compiler/z/runtime/PicBuilder.m4:171]: We should try to keep the names of these the same as the fields for easier searchability. For example searching for "totalSizeOfPicSlots" will not get a hit here.

**Comment by fjeremic** [comment]
I still need to review PicBuilder. Will do that in another round.

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp (hunk 1 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/codegen/S390J9CallSnippet.cpp (hunk 2 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390J9CallSnippet.cpp]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 3 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/runtime/PicBuilder.m4 (hunk 7 above)
Comment (r30shah) [reply runtime/compiler/z/runtime/PicBuilder.m4]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 5 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: postDeps is a copy of existing `dependencies` to make room for new post dependencies I need for this function. Actually I can remove those extra lines of code as I have done in https://github.com/eclipse/openj9/compare/3282a01df297676d33999fca192d2613b5db1b09..d1f154c59f4e002f94b5e613001eb016a95a6927

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 3 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 5 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: It is actually not needed, I wanted to just extract the post deps from the original dependencies and add room for extra deps I need in this function. There is a better way to achieve that and I have made change in https://github.com/eclipse/openj9/compare/3282a01df297676d33999fca192d2613b5db1b09..d1f154c59f4e002f94b5e613001eb016a95a6927

File: runtime/compiler/z/runtime/PicBuilder.m4 (hunk 7 above)
Comment (r30shah) [reply runtime/compiler/z/runtime/PicBuilder.m4]: Fixed in https://github.com/eclipse/openj9/compare/ab82c5d254e0315b3c65722371fda75cc1d79a10..3282a01df297676d33999fca192d2613b5db1b09

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 6 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1781]: According to comment in https://github.com/eclipse/openj9/blob/67d8cf0cd31d5215756678a1cd434fc04af98055/runtime/compiler/z/codegen/S390PrivateLinkage.cpp#L2189 It is needed for the patchupVirtualCall, I feel it needs more clear documentation then that. I remember, while testing when I commented out the NOP it was failing the tests, let me find-out the exact call path that that requires this NOP.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 4 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1695]: How about using a macro `NUMBER_OF_PIC_SLOTS_FOR_RX` instead of adding additional field in ccInterfaceData? This CCData is allocated per call site and we will have additional field added to all the call site's CCInterfaceData with information which actually is a constant (ATM number of PIC slots are constant and can not be changed at compilation time).

Comment (fjeremic) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1695]: `static const` is allocated once for the entire DLL, not per call site. It is a static variable after all. It will not incur any difference to `sizeof(ccInterfaceData)`.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 6 above)
Comment (fjeremic) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1781]: Ok, we should put the same comment here as exists in the JVM path linked above. I could definitely see someone not knowing anything about this coming down and removing the NOP thinking it shouldn't exist.

Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1782]: I have copied the same comment in https://github.com/eclipse/openj9/compare/e2eb175cf4e392b9a013648129c1209bc1790bc8..7e34301ef20a2e2b9b78a0d987f8c0e5eb801d0c 
Currently checking the path that needs this NOP, will modify if I find something useful.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 4 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Sorry, over sighted the suggestion and forgot we can have static members inside struct in C++. Added the static const inside the ccInterfaceData struct to reflect number of PIC slots, https://github.com/eclipse/openj9/compare/e2eb175cf4e392b9a013648129c1209bc1790bc8..7e34301ef20a2e2b9b78a0d987f8c0e5eb801d0c.

**Review by fjeremic — APPROVED** [review]
